contracts/extensions/VotingReputation.sol

```diff
@@ -0,0 +1,893 @@
+  /// @param _action A bytes array encoding a function call
+  /// @param _key Reputation tree key for the root domain
+  /// @param _value Reputation tree value for the root domain
+  /// @param _branchMask The branchmask of the proof
+  /// @param _siblings The siblings of the proof
+  function createRootPoll(
+    address _target,
+    bytes memory _action,
+    bytes memory _key,
+    bytes memory _value,
+    uint256 _branchMask,
+    bytes32[] memory _siblings
+  )
+    public
+  {
+    uint256 rootSkillId = colony.getDomain(1).skillId;
+    createPoll(_target, _action, 1, rootSkillId, _key, _value, _branchMask, _siblings);
+  }
+
+  /// @notice Create a poll in any domain
+  /// @param _domainId The domain where we vote on the poll
+  /// @param _childSkillIndex The childSkillIndex pointing to the domain of the action
+  /// @param _target The contract to which we send the action (0x0 for the colony)
+  /// @param _action A bytes array encoding a function call
+  /// @param _key Reputation tree key for the domain
+  /// @param _value Reputation tree value for the domain
+  /// @param _branchMask The branchmask of the proof
+  /// @param _siblings The siblings of the proof
+  function createDomainPoll(
+    uint256 _domainId,
+    uint256 _childSkillIndex,
+    address _target,
+    bytes memory _action,
+    bytes memory _key,
+    bytes memory _value,
+    uint256 _branchMask,
+    bytes32[] memory _siblings
+  )
+    public
+  {
+    uint256 domainSkillId = colony.getDomain(_domainId).skillId;
+    uint256 actionDomainSkillId = getActionDomainSkillId(_action);
+
+    if (domainSkillId != actionDomainSkillId) {
+      uint256 childSkillId = colonyNetwork.getChildSkillId(domainSkillId, _childSkillIndex);
+      require(childSkillId == actionDomainSkillId, "voting-rep-invalid-domain-id");
+    }
+
+    createPoll(_target, _action, _domainId, domainSkillId, _key, _value, _branchMask, _siblings);
+  }
+
+  /// @notice Stake on a poll
+  /// @param _pollId The id of the poll
+  /// @param _permissionDomainId The domain where the extension has the arbitration permission
+  /// @param _childSkillIndex For the domain in which the poll is occurring
+  /// @param _vote The side being supported (0 = NAY, 1 = YAY)
+  /// @param _amount The amount of tokens being staked
+  /// @param _key Reputation tree key for the staker/domain
+  /// @param _value Reputation tree value for the staker/domain
+  /// @param _branchMask The branchmask of the proof
+  /// @param _siblings The siblings of the proof
+  function stakePoll(
+    uint256 _pollId,
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _vote,
+    uint256 _amount,
+    bytes memory _key,
+    bytes memory _value,
+    uint256 _branchMask,
+    bytes32[] memory _siblings
+  )
+    public
+  {
+    Poll storage poll = polls[_pollId];
+    require(_vote <= 1, "voting-rep-bad-vote");
+    require(_amount > 0, "voting-rep-bad-amount");
```

Comment:
While it is what I said needed to be checked, I was wrong (again caught out by `_amount` vs `amount`) and this isn't actually sufficient to prevent what I was talking about - `_amount` can be greater than zero to pass this check, and `amount` will be set to zero by the `min` if `YAY` has been staked fully.

We also need to consider what happen in the case where a poll is escalated to a domain that has the same reputation distribution, which means `poll.stakes[YAY]` and `poll.stakes[NAY]` are fully staked immediately on escalation. I *think* everything mostly works, but it looks like you can end up with `poll.events[STAKE_END]` being set to be after `poll.events[SUBMIT_END]`, and not fire the `PollEventSet(_pollId, STAKE_END)` event which could confuse clients.

Reply:
Mm ok let's make the require for `amount` then. And regarding the second point... I suppose we could add a check during `escalatePoll` and update the timestamp / fire event as needed.